# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/style/SharedSubResourceCache.h
# Commit: 313c1f782833
# Full Hash: 313c1f7828336b6163271454aaface96f0e3fc96
# Author: Tooru Fujisawa <arai_a@mac.com>
# Date: 2024-10-02 09:50:09
# Regressor Bug: 1899734
# File Overlap Count: 1
# Description:
#   Bug 1899734 - Part 5: Make SharedSubResourceCache capable of handling CacheablePerformanceTimingData. r=emilio
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D223351
# ==============================================================================

diff -r baa464d5f255 -r 313c1f782833 layout/style/SharedSubResourceCache.h
--- a/layout/style/SharedSubResourceCache.h	Tue Oct 01 12:17:21 2024 +0000
+++ b/layout/style/SharedSubResourceCache.h	Tue Oct 01 12:17:22 2024 +0000
@@ -27,6 +27,7 @@
 //   SheetLoadData.
 
 #include "mozilla/PrincipalHashKey.h"
+#include "mozilla/RefPtr.h"
 #include "mozilla/WeakPtr.h"
 #include "nsTHashMap.h"
 #include "nsIMemoryReporter.h"
@@ -36,10 +37,42 @@
 #include "mozilla/dom/CacheExpirationTime.h"
 #include "mozilla/dom/Document.h"
 #include "nsContentUtils.h"
+#include "nsISupportsImpl.h"
 #include "mozilla/StaticPtr.h"
+#include "mozilla/dom/CacheablePerformanceTimingData.h"
 
 namespace mozilla {
 
+// A struct to hold the network-related metadata associated with the cache.
+//
+// When inserting a cache, the consumer should create this from the request and
+// make it available via
+// SharedSubResourceCacheLoadingValueBase::GetNetworkMetadata.
+//
+// When using a cache, the consumer can retrieve this from
+// SharedSubResourceCache::Result::mNetworkMetadata and use it for notifying
+// the observers once the necessary data becomes ready.
+// This struct is ref-counted in order to allow this usage.
+class SubResourceNetworkMetadataHolder {
+ public:
+  SubResourceNetworkMetadataHolder() = delete;
+
+  explicit SubResourceNetworkMetadataHolder(nsIRequest* aRequest);
+
+  const dom::CacheablePerformanceTimingData* GetPerfData() const {
+    return mPerfData.ptrOr(nullptr);
+  }
+
+  NS_INLINE_DECL_THREADSAFE_REFCOUNTING(SubResourceNetworkMetadataHolder)
+
+ private:
+  ~SubResourceNetworkMetadataHolder() = default;
+
+  mozilla::Maybe<dom::CacheablePerformanceTimingData> mPerfData;
+
+  // TODO: Add HTTP headers for DevTools/WebDriver notifications (bug 1915626).
+};
+
 enum class CachedSubResourceState {
   Miss,
   Loading,
@@ -56,6 +89,8 @@
   virtual bool IsCancelled() const = 0;
   virtual bool IsSyncLoad() const = 0;
 
+  virtual SubResourceNetworkMetadataHolder* GetNetworkMetadata() const = 0;
+
   virtual void StartLoading() = 0;
   virtual void SetLoadCompleted() = 0;
   virtual void OnCoalescedTo(const Derived& aExistingLoad) = 0;
@@ -113,11 +148,13 @@
  protected:
   struct CompleteSubResource {
     RefPtr<Value> mResource;
+    RefPtr<SubResourceNetworkMetadataHolder> mNetworkMetadata;
     CacheExpirationTime mExpirationTime = CacheExpirationTime::Never();
     bool mWasSyncLoad = false;
 
     explicit CompleteSubResource(LoadingValue& aValue)
         : mResource(aValue.ValueForCache()),
+          mNetworkMetadata(aValue.GetNetworkMetadata()),
           mExpirationTime(aValue.ExpirationTime()),
           mWasSyncLoad(aValue.IsSyncLoad()) {}
 
@@ -127,6 +164,8 @@
  public:
   struct Result {
     Value* mCompleteValue = nullptr;
+    RefPtr<SubResourceNetworkMetadataHolder> mNetworkMetadata;
+
     LoadingValue* mLoadingOrPendingValue = nullptr;
     CachedSubResourceState mState = CachedSubResourceState::Miss;
 
@@ -134,6 +173,7 @@
 
     explicit constexpr Result(const CompleteSubResource& aCompleteSubResource)
         : mCompleteValue(aCompleteSubResource.mResource.get()),
+          mNetworkMetadata(aCompleteSubResource.mNetworkMetadata),
           mLoadingOrPendingValue(nullptr),
           mState(CachedSubResourceState::Complete) {}
 